### PR TITLE
Grpc tls auth

### PR DESCRIPTION
**Pull request by dschowta**
Fixes to provide GRPC based authentication

### CODE REVIEW

**Comment by dschowta** [comment]
TODO: updating  apidoc

File: main.go [hunk 1]
```diff
@@ -225,20 +260,165 @@ func main() {
...
+	if pkiConf.ServerKey == "" || pkiConf.ServerCert == "" || pkiConf.CaCert == "" {
+		log.Printf("In order to run GRPC server, ServerKey file, Server Cert file and CA Cert file must be set in conf.pki setting")
+		return fmt.Errorf("Certificate and key files are not set")
+	}
+	if fileExists(pkiConf.ServerKey) && fileExists(pkiConf.ServerCert) && fileExists(pkiConf.CaCert) {
+		log.Print("Using the existing server certificates")
+		return nil
+	}
+	if !fileExists(pkiConf.ServerCert) && ca == nil {
+		return fmt.Errorf("Server certificate does not exist CA is not set up to sign the server")
```
Comment (farshidtz) [inline main.go:273]: error message is not clear.

File: main.go [hunk 2]
```diff
@@ -225,20 +260,165 @@ func main() {
...
+		if err != nil {
+			return nil, fmt.Errorf("error writing the CA Certificate file: %v", err)
+		}
+		err = ioutil.WriteFile(pkiConf.CaKey, key, 0600)
+		if err != nil {
+			return nil, fmt.Errorf("error writing the CA key file: %v", err)
+		}
+	} else {
+		log.Printf("reusing the existing CA: %s", pkiConf.CaCert)
+		ca, err = pki.NewCAFromFile(pkiConf.CaCert, pkiConf.CaKey)
```
Comment (farshidtz) [inline main.go:369]: return error with appropriate prefix.

File: main.go [hunk 3]
```diff
@@ -225,20 +260,165 @@ func main() {
...
+	certificate, err := tls.LoadX509KeyPair(serverCertFile, serverPrivatekey)
+	if err != nil {
+		log.Panicf("could not load server key pair: %s", err)
+	}
+
+	// Create a certificate pool from the certificate authority
+	certPool := x509.NewCertPool()
+	ca, err := ioutil.ReadFile(caFile)
+	if err != nil {
+		log.Fatalf("could not read ca certificate: %s", err)
```
Comment (farshidtz) [inline main.go:391]: use panic?

**Review by farshidtz — APPROVED** [review]

**Pull request merged**